crates/matrix-sdk/src/room/timeline/event_handler.rs

```diff
@@ -0,0 +1,460 @@
+    raw: Raw<AnySyncTimelineEvent>,
+    encryption_info: Option<EncryptionInfo>,
+    own_user_id: &UserId,
+    position: EventPosition,
+    timeline: &TimelineInner,
+) {
+    let event = match raw.deserialize() {
+        Ok(ev) => ev,
+        Err(_e) => {
+            // TODO: Add some sort of error timeline item
+            return;
+        }
+    };
+
+    let sender = event.sender().to_owned();
+    let is_own_event = sender == own_user_id;
+
+    let meta = EventMetadata {
+        raw_event: Some(raw),
+        sender,
+        origin_server_ts: Some(event.origin_server_ts()),
+        is_own_event,
+        relations: event.relations().cloned(),
+        encryption_info,
+    };
+    let flow = Flow::Remote {
+        event_id: event.event_id().to_owned(),
+        txn_id: event.transaction_id().map(ToOwned::to_owned),
+        position,
+    };
+
+    EventHandler::new(meta, flow, timeline).handle_event(event.into())
+}
+
+enum Flow {
+    Local { txn_id: OwnedTransactionId },
+    Remote { event_id: OwnedEventId, txn_id: Option<OwnedTransactionId>, position: EventPosition },
+}
+
+impl Flow {
+    fn to_key(&self) -> TimelineKey {
+        match self {
+            Self::Remote { event_id, .. } => TimelineKey::EventId(event_id.to_owned()),
+            Self::Local { txn_id } => TimelineKey::TransactionId(txn_id.to_owned()),
+        }
+    }
+}
+
+struct EventMetadata {
+    raw_event: Option<Raw<AnySyncTimelineEvent>>,
+    sender: OwnedUserId,
+    origin_server_ts: Option<MilliSecondsSinceUnixEpoch>,
+    is_own_event: bool,
+    relations: Option<Relations>,
+    encryption_info: Option<EncryptionInfo>,
+}
+
+impl From<AnySyncTimelineEvent> for EventKind {
+    fn from(event: AnySyncTimelineEvent) -> Self {
+        match event {
+            AnySyncTimelineEvent::MessageLike(AnySyncMessageLikeEvent::RoomRedaction(
+                SyncRoomRedactionEvent::Original(OriginalSyncRoomRedactionEvent {
+                    redacts,
+                    content,
+                    ..
+                }),
+            )) => Self::Redaction { redacts, content },
+            AnySyncTimelineEvent::MessageLike(ev) => match ev.original_content() {
+                Some(content) => Self::Message { content },
+                None => Self::RedactedMessage,
+            },
+            AnySyncTimelineEvent::State(ev) => match ev.original_content() {
+                Some(_content) => Self::State { _content },
+                None => Self::RedactedState,
+            },
+        }
+    }
+}
+
+#[derive(Clone)]
+enum EventKind {
+    Message { content: AnyMessageLikeEventContent },
+    RedactedMessage,
+    Redaction { redacts: OwnedEventId, content: RoomRedactionEventContent },
+    // FIXME: Split further for state keys of different type
+    State { _content: AnyStateEventContent },
+    RedactedState, // AnyRedactedStateEventContent
+}
+
+enum EventPosition {
+    Start,
+    End,
+}
+
+struct EventHandler<'a> {
```

Comment:
Oh no, now we have two things called `EventHandler` in the codebase? Is there a reason for this to exist except to keep the number of methods on the `TimelineInner` struct slim?